### PR TITLE
Fixed electricity meter handling withe dual import and export rates

### PR DESCRIPTION
**Pull request by 696GrocuttT**
Multiple MPAN numbers can belong to a single meter serial number, eg if separant
export readings are used due to solar panels. This patch adds the MPAN number to
the various electricity sensors to avoid duplicates with name clashes. This patch
also adds support for a separate tariff for each MPAN, to properly support
different import and export rates.

### CODE REVIEW

**Comment by BottlecapDave** [comment]
Hello. It looks like there's a few issues with your PR. The limited automated tests I have aren't passing. 

These can be run `API_KEY=<<OCTOPUS_API_KEY>> python -m pytest tests`

Also when I pulled your branch and installed it on my machine wth some target sensors, I'm getting the following error

```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 431, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 718, in async_device_update
    await task
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 336, in async_update
    await self.coordinator.async_request_refresh()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 141, in async_request_refresh
    await self._debounced_refresh.async_call()
  File "/usr/src/homeassistant/homeassistant/helpers/debounce.py", line 78, in async_call
    await task
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 165, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 265, in _async_refresh
    update_callback()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 325, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 539, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 572, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in _stringify_state
    if (state := self.state) is None:
  File "/usr/src/homeassistant/homeassistant/components/binary_sensor/__init__.py", line 209, in state
    if (is_on := self.is_on) is None:
  File "/config/custom_components/octopus_energy/binary_sensor.py", line 98, in is_on
    self._target_rates = self.calculate_continuous_times()
  File "/config/custom_components/octopus_energy/binary_sensor.py", line 169, in calculate_continuous_times
    rates = self.get_applicable_rates()
  File "/config/custom_components/octopus_energy/binary_sensor.py", line 158, in get_applicable_rates
    rateData = list(test.values())[0]
NameError: name 'test' is not defined
```

File: custom_components/octopus_energy/binary_sensor.py [hunk 1]
```diff
@@ -152,9 +153,15 @@ def get_applicable_rates(self):
     # Retrieve the rates that are applicable for our target rate
     rates = []
     if self.coordinator.data != None:
-      for rate in self.coordinator.data:
-        if rate["valid_from"] >= target_start and (target_end == None or rate["valid_to"] <= target_end):
-          rates.append(rate)
+      ratesData = self.coordinator.data
+      if len(ratesData) == 1:
+        rateData = list(test.values())[0]
```
Comment (BottlecapDave) [inline custom_components/octopus_energy/binary_sensor.py:158]: Based on my testing, I think this line needs to be

```
rateData = next(iter(ratesData.values()))
```

**Comment by BottlecapDave** [comment]
Looks like the tests are working locally, so the tests are probably being run in your environment/fork which might not have `API_KEY` set. It looks like everything is coming back as expected in my local HA instance, once you make the specified change in the binary sensor.

**Comment by 696GrocuttT** [comment]
Oops, sorry about that. It looks like that line wasn't being executed in my setup because I have multiple rates. I've fixed that line so hopefully it'll be ok now. 

I run homeassistant from the supported docker image, now from the raw python venv. As a result I wasn't able to get the tests running for other reasons.

**Comment by BottlecapDave** [comment]
No problem. Thanks for taking the time to make this improvement.

**Pull request merged**